# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Importing Data

In [2]:
path=r"C:\Users\chenh\z_Careerfoundry\A4_Instacart Basket Analysis"

df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

# If-Statements with User-Defined Functions

In [3]:
# Create a subset
df_merged_part=df_ords_prods_merge[:100000]

In [4]:
# Define a function for price labels

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [5]:
# Apply price range function.
df_merged_part['price_range']=df_merged_part.apply(price_label,axis=1)

C:\Users\chenh\AppData\Local\Temp/ipykernel_17912/3156898998.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_merged_part['price_range']=df_merged_part.apply(price_label,axis=1)


In [6]:
df_merged_part['price_range'].value_counts(dropna=False)

Mid-range product    87841
Low-range product    12159
Name: price_range, dtype: int64

# If-Statements with the loc() Function

In [7]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices']>15,'price_range_loc'] = 'High-range product'

In [8]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices']>5) &(df_ords_prods_merge['prices']<=15),'price_range_loc']= 'Mid-range product'

In [9]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [10]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

Mid-range product     21865987
Low-range product     10126321
High-range product      412551
Name: price_range_loc, dtype: int64

# If-Statements with For-Loops

In [11]:
# Find frequency of orders_day_of_week.
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [12]:
result = []

for value in df_ords_prods_merge['orders_day_of_week']:
    if value==0:
        result.append('Busiest day')
    elif value==4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [13]:
# Create new column in ords_prods_merge to view results in context.
df_ords_prods_merge['busiest_day'] = result

In [14]:
df_ords_prods_merge['busiest_day'].value_counts(dropna=False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

# Task

## STEP2

In [15]:
# Create column “busiest_days" from "orders_day_of_week" to identify two busiest days (0, 1), slowest days (4, 3), or regularly busy (all other days).

result_2=[]
for value in df_ords_prods_merge['orders_day_of_week']:
    if value==0 or value==1:
        result_2.append('Busiest days')
    elif value == 4 or value == 3:
        result_2.append("Slowest days")
    else:
        result_2.append("Regularly busy")

In [16]:
# Create new 'busiest_days' column in ords_prods_merge to view results in context.

df_ords_prods_merge['busiest_days'] = result_2

In [17]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

## STEP3

Total "Busiest days" = 11864412, which matches orders_day_of_week 0 + 1, or 6204182 + 5660230).
Total "Slowest days" = 7624336, which matches orders_day_of_week 4 + 3, or 3783802 + 3840534).

## STEP4

In [18]:
# Check value counts in "order_hour_of_day" column.
df_ords_prods_merge['order_hour_of_day'].value_counts()

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

The value counts listed above will be split into equal thirds for the following labels: "Most orders" = 10, 11, 14, 15, 13, 12, 16, 9; "Fewest orders" = 23, 6, 0, 1, 5, 2, 4, and 3; "Average orders" = all remaining values in order_hour_of_day.

In [19]:
# Create for-loop if statement labeling periods of time as “Most orders,” “Average orders,” and “Fewest orders.”

result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23,6,0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [20]:
# Create new column "busiest_period_of_day" in ords_prods_merge.

df_ords_prods_merge['busiest_period_of_day'] = result_3

## STEP5

In [21]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

# Export Dataframe

In [22]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_derived.pkl'))